File: python/nqueen.py
class NQueensSolver:
    def __init__(self, n):
        self.n = n
        self.board = [["." for _ in range(n)] for _ in range(n)]
        self.solutions = []

    def solve(self):
        self.backtrack(0)
        return self.solutions

    def is_safe(self, row, col):
        # Check if there is a queen in the same column
        for i in range(row):
            if self.board[i][col] == "Q":
                return False

        # Check upper left diagonal
        i, j = row, col
        while i >= 0 and j >= 0:
            if self.board[i][j] == "Q":
                return False
            i -= 1
            j -= 1

        # Check upper right diagonal
        i, j = row, col
        while i >= 0 and j < self.n:
            if self.board[i][j] == "Q":
                return False
            i -= 1
            j += 1

        return True

    def backtrack(self, row):
        if row == self.n:
            solution = ["".join(row) for row in self.board]
            self.solutions.append(solution)
            return

        for col in range(self.n):
            if self.is_safe(row, col):
                self.board[row][col] = "Q"
                self.backtrack(row + 1)
                self.board[row][col] = "."


# n = 4
n = int(input("Enter the value of n:"))
solver = NQueensSolver(n)
solutions = solver.solve()
for index, solution in enumerate(solutions):
    print(f"Solution {index + 1}:")
    for row in solution:
        print(row)
    print()

print(f"Found {len(solutions)} solutions for {n}-queens problem:")

File: javascript/nqueen.js
class NQueensSolver {
  constructor(n) {
    this.n = n;
    this.board = [];
    for (let i = 0; i < n; i++) {
      this.board[i] = [];
      for (let j = 0; j < n; j++) {
        this.board[i][j] = ".";
      }
    }
    this.solutions = [];
  }

  solve() {
    this.backtrack(0);
    return this.solutions;
  }

  isSafe(row, col) {
    // Check if there is a queen in the same column
    for (let i = 0; i < row; i++) {
      if (this.board[i][col] === "Q") {
        return false;
      }
    }

    // Check upper left diagonal
    for (let i = row, j = col; i >= 0 && j >= 0; i--, j--) {
      if (this.board[i][j] === "Q") {
        return false;
      }
    }

    // Check upper right diagonal
    for (let i = row, j = col; i >= 0 && j < this.n; i--, j++) {
      if (this.board[i][j] === "Q") {
        return false;
      }
    }

    return true;
  }

  backtrack(row) {
    if (row === this.n) {
      const solution = [];
      for (let i = 0; i < this.n; i++) {
        solution.push(this.board[i].join(""));
      }
      this.solutions.push(solution);
      return;
    }

    for (let col = 0; col < this.n; col++) {
      if (this.isSafe(row, col)) {
        this.board[row][col] = "Q";
        this.backtrack(row + 1);
        this.board[row][col] = ".";
      }
    }
  }
}

// const n = 4;
const n = +prompt("Enter the value of n:");
const solver = new NQueensSolver(n);
const solutions = solver.solve();
console.log(`Found ${solutions.length} solutions for ${n}-queens problem:`);
solutions.forEach((solution, index) => {
  console.log(`Solution ${index + 1}:`);
  solution.forEach((row) => {
    console.log(row);
  });
  console.log();
});
